File: sandbox-runner/patterns/mapreduce-tree/typescript/workflows.ts
import {
  condition,
  defineSignal,
  getExternalWorkflowHandle,
  log,
  proxyActivities,
  setHandler,
  startChild,
  workflowInfo,
} from "@temporalio/workflow";

import type * as activities from "./activities";
import { LEAF_THRESHOLD, MAX_DEPTH, RESULT_SIGNAL, TASK_QUEUE } from "./shared";

const { processLeaf } = proxyActivities<typeof activities>({
  startToCloseTimeout: "30 seconds",
});

/** Signal payload sent from child to parent. */
export interface ResultPayload {
  id: string;
  results: string[];
}

export const resultSignal = defineSignal<[ResultPayload]>(RESULT_SIGNAL);

/**
 * Leaf workflow: processes one record via an Activity and signals the result
 * back to the parent Node workflow.
 */
export async function leafWorkflow(record: string, parentWorkflowId: string): Promise<void> {
  const result = await processLeaf(record);
  log.info(`Leaf processed`, { record, result });

  const parent = getExternalWorkflowHandle(parentWorkflowId);
  await parent.signal(resultSignal, { id: record, results: [result] });
}

/**
 * Node workflow: recursively splits the record set or fans out to leaf workflows,
 * collects results via signals, then signals its own result up to its parent.
 */
export async function nodeWorkflow(
  records: string[],
  depth: number = 0,
  parentWorkflowId: string = ""
): Promise<string[]> {
  if (depth > MAX_DEPTH) {
    throw new Error(`Tree depth exceeded ${MAX_DEPTH}`);
  }

  const myId = workflowInfo().workflowId;
  const collectedResults: string[] = [];
  let received = 0;
  let expected = 0;

  setHandler(resultSignal, (payload: ResultPayload) => {
    collectedResults.push(...payload.results);
    received++;
  });

  if (records.length <= LEAF_THRESHOLD) {
    // Fan out to leaf workflows — one per record.
    expected = records.length;
    for (const record of records) {
      void startChild(leafWorkflow, {
        args: [record, myId],
        workflowId: `${myId}/leaf-${record}`,
        taskQueue: TASK_QUEUE,
      });
    }
  } else {
    // Split and recurse into child node workflows.
    const mid = Math.floor(records.length / 2);
    const chunks = [records.slice(0, mid), records.slice(mid)];
    expected = chunks.length;
    for (let i = 0; i < chunks.length; i++) {
      void startChild(nodeWorkflow, {
        args: [chunks[i], depth + 1, myId],
        workflowId: `${myId}/node-d${depth + 1}-${i}`,
        taskQueue: TASK_QUEUE,
      });
    }
  }

  // Wait until all expected signals have arrived.
  await condition(() => received >= expected);

  log.info(`Node complete`, { depth, records: records.length, results: collectedResults.length });

  // Signal aggregated results up to parent (if this is not the root).
  if (parentWorkflowId) {
    const parent = getExternalWorkflowHandle(parentWorkflowId);
    await parent.signal(resultSignal, { id: myId, results: collectedResults });
  }

  return collectedResults;
}

File: sandbox-runner/patterns/mapreduce-tree/python/workflows.py
from dataclasses import dataclass
from datetime import timedelta

from temporalio import workflow
from temporalio.workflow import ChildWorkflowHandle

from activities import process_leaf
from shared import LEAF_THRESHOLD, MAX_DEPTH, RESULT_SIGNAL, TASK_QUEUE


@dataclass
class ResultPayload:
    id: str
    results: list[str]


@workflow.defn
class LeafWorkflow:
    """Leaf workflow: processes one record and signals the result back to parent."""

    @workflow.run
    async def run(self, record: str, parent_workflow_id: str) -> None:
        result = await workflow.execute_activity(
            process_leaf,
            record,
            start_to_close_timeout=timedelta(seconds=30),
        )
        workflow.logger.info(f"Leaf processed: {record} → {result}")

        parent = workflow.get_external_workflow_handle(parent_workflow_id)
        await parent.signal(RESULT_SIGNAL, ResultPayload(id=record, results=[result]))


@workflow.defn
class NodeWorkflow:
    """Node workflow: splits or fans out to leaves, collects results via signals,
    and signals aggregated results up to its parent."""

    def __init__(self) -> None:
        self._collected: list[str] = []
        self._received = 0

    @workflow.signal(name=RESULT_SIGNAL)
    def node_result(self, payload: ResultPayload) -> None:
        self._collected.extend(payload.results)
        self._received += 1

    @workflow.run
    async def run(
        self,
        records: list[str],
        depth: int = 0,
        parent_workflow_id: str = "",
    ) -> list[str]:
        if depth > MAX_DEPTH:
            raise RuntimeError(f"Tree depth exceeded {MAX_DEPTH}")

        my_id = workflow.info().workflow_id
        expected = 0

        if len(records) <= LEAF_THRESHOLD:
            # Fan out to leaf workflows — one per record.
            expected = len(records)
            for record in records:
                await workflow.start_child_workflow(
                    LeafWorkflow.run,
                    args=[record, my_id],
                    id=f"{my_id}/leaf-{record}",
                    task_queue=TASK_QUEUE,
                )
        else:
            # Split and recurse into child node workflows.
            mid = len(records) // 2
            chunks = [records[:mid], records[mid:]]
            expected = len(chunks)
            for i, chunk in enumerate(chunks):
                await workflow.start_child_workflow(
                    NodeWorkflow.run,
                    args=[chunk, depth + 1, my_id],
                    id=f"{my_id}/node-d{depth+1}-{i}",
                    task_queue=TASK_QUEUE,
                )

        await workflow.wait_condition(lambda: self._received >= expected)

        workflow.logger.info(
            f"Node complete: depth={depth} records={len(records)} results={len(self._collected)}"
        )

        if parent_workflow_id:
            parent = workflow.get_external_workflow_handle(parent_workflow_id)
            await parent.signal(
                RESULT_SIGNAL, ResultPayload(id=my_id, results=self._collected)
            )

        return self._collected
